### PR TITLE
get border, bug fixes, optimze

### PR DESCRIPTION
**Pull request by SkrighYZ**
1. Scout: fix bug when calculating enemy distance 
2. Scout: rather than estimating clusters of troops once for each scout, only calculate once each turn (doesnt make too much difference if using random sampling rather than kmeans for estimation)
3. Border: add get_border() to calculate the territory border at any time
4. Minor change: change self.map_states to np array, add 1e-7 when dividing force magnitude for stability

### CODE REVIEW

**Review by kybli — APPROVED** [review]
LGTM!

File: players/g3_player.py [hunk 1]
```diff
@@ -207,51 +208,107 @@ def play(self, unit_id: List[List[str]], unit_pos: List[List[Point]], map_states
...
         ally_units = np.delete(self.our_units, scout_ids, axis=0)
         ally_dist = ((scout_units.reshape(-1, 1, 2) - ally_units.reshape(1, -1, 2)) ** 2).sum(axis=2)
         min_ally_id = ally_dist.argmin(axis=1)
-        min_enemy_dist = ((scout_units.reshape(-1, 1, 2) - self.enemy_units.reshape(1, -1, 2)) ** 2).sum(axis=2).argmin(axis=1)
+        min_enemy_dist = ((scout_units.reshape(-1, 1, 2) - self.enemy_units.reshape(1, -1, 2)) ** 2).sum(axis=2).min(axis=1)
+
+        enemy_clusters, ally_clusters = self._get_clusters(ally_units)
         for i in range(scout_units.shape[0]):
-            if ally_dist[i, min_ally_id[i]] >= min_enemy_dist[i]:
+            if ally_dist[i, min_ally_id[i]] >= min_enemy_dist[i] * 2:
```
Comment (yunlanli) [inline players/g3_player.py:305]: curious why a value of 2 is used.

File: players/g3_player.py [hunk 2]
```diff
@@ -207,51 +208,107 @@ def play(self, unit_id: List[List[str]], unit_pos: List[List[Point]], map_states
...
+        if self.us < 2: # 0, 1
+            for i in range(100):
+                if self.map_states[i, 99*self.us] != self.us:
+                    pt = (i-1, 99*self.us)
+                    break
+        else: # 2, 3
+            for i in range(100):
+                if self.map_states[99-i, 99*(3-self.us)] != self.us:
+                    pt = (99-i+1, 99*(3-self.us))
+                    break
```
Comment (yunlanli) [inline players/g3_player.py:240]: If I understand correctly, you are trying to find the map edge that are part of a player's border here.

It seems that here player index 1 (i.e player 2) is at the top right, but I think the simulator puts player on bottom left?

File: players/g3_player.py (hunk 1 above)
Comment (SkrighYZ) [reply players/g3_player.py:305]: No specific reason actually. 1 is guaranteed to be safe (all directions considered; e.g., worst case enemy and ally come from exact same direction) but to be more aggressive 2 seems to work fine.

File: players/g3_player.py (hunk 2 above)
Comment (SkrighYZ) [reply players/g3_player.py:240]: (0, 99) is indeed bottom left...

File: players/g3_player.py [hunk 3]
```diff
@@ -207,51 +208,107 @@ def play(self, unit_id: List[List[str]], unit_pos: List[List[Point]], map_states
...
+        # currently ignoring disputed cells
+        if self.map_states[pt] != self.us: 
+            return False
+
+        xmax, ymax = self.map_states.shape
+        neighbors = np.array([[min(pt[0]+1, xmax-1), pt[1]], 
+            [max(pt[0]-1, 0), pt[1]],
+            [pt[0], min(pt[1]+1, ymax-1)],
+            [pt[0], max(pt[1]-1, 0)]])
+        return any(self.map_states[neighbors[:, 0], neighbors[:, 1]] != self.us)
```
Comment (yunlanli) [inline players/g3_player.py:273]: looks good to me!

**Review by yunlanli — APPROVED** [review]

File: players/g3_player.py (hunk 2 above)
Comment (yunlanli) [reply players/g3_player.py:240]: gotcha, I think I confused the position of the element in the matrix with where it is on the map.

Comment (SkrighYZ) [reply players/g3_player.py:240]: When i wrote the code, I still think about it with the matrix and just flip player index 1 and 3 in my head. Doesn't change anything, unless moves are involved

**Pull request merged**